### PR TITLE
closing bug 5315

### PR DESCRIPTION
**Pull request by forestjulien**
This patch closes [bug 5315](https://coq.inria.fr/bugs/show_bug.cgi?id=5315) and an unreported bug on Function.
@mattam82 @Matafou @maximedenes: The most important change (outside of Function)  is the change in pretyping which is needed in order to typecheck term like f (fun x => ...) for which I need to know the type of x at glob_term level. 
The modification is triggered by Flags.is_function_mode which is itself set to true only during the hole resolution phase of Function so I can see any possible change outside of Function itself in pretyping. 

Any comments is welcome.

### CODE REVIEW

**Comment by ejgallego** [comment]
Hi @forestjulien , I am afraid that we are trying to make parameters to functions more local, so I'd suggest you move the flag from the global file to the pretyper itself.

File: pretyping/pretyping.ml
Comment (ejgallego) [inline pretyping/pretyping.ml]: please don't use a wildcard here.

File: lib/flags.ml
Comment (ejgallego) [inline lib/flags.ml]: Please avoid to introduce global flags.

Comment (forestjulien) [inline lib/flags.ml]: What do you prefer: 
I can move that code in Pretyping module and export it to the whole word or add an optional argument to the full chain from pretype_type to ise_pretype_gen. 
I must admit I would prefer the first one (less code to write) but I can do both

Comment (ejgallego) [inline lib/flags.ml]: Let's see what @ppedrot has to say here.

Comment (forestjulien) [inline lib/flags.ml]: Ok

Comment (forestjulien) [inline lib/flags.ml]: For now, in order to gain time I've push the first sol

File: pretyping/pretyping.ml
Comment (forestjulien) [inline pretyping/pretyping.ml]: @ejgallego: Is that ok for you ?

Comment (forestjulien) [inline pretyping/pretyping.ml]: Do not answer, I've found a much better solution

**Comment by forestjulien** [comment]
The failure doe not seems to come from the patch itself as long as I understand it

**Comment by Zimmi48** [comment]
@forestjulien You are right. The fiat crypto target in Travis is currently broken in v8.6. A fix is awaiting in #792.

**Comment by Matafou** [comment]
Nice solution @forestjulien. This version can be merged I guess since it does not touch anything else than the indfun plugin.

**Comment by ejgallego** [comment]
Sorry, the solution currently introduces a global flag in the pretyper so IMHO it needs more discussion.

**Comment by Matafou** [comment]
Looking at it closer maybe you can rebase into one commit and clean a bit your patch before it is merged @forestjulien. Some comments were re-formatted uselessly, some newlines were added too.

File: lib/flags.mli
Comment (Matafou) [inline lib/flags.mli]: clean please

File: pretyping/pretyping.ml
Comment (Matafou) [inline pretyping/pretyping.ml]: Set Function does not exist. Please remove reference to it.

File: test-suite/bugs/closed/5315.v
Comment (Matafou) [inline test-suite/bugs/closed/5315.v]: Should this comment be removed?

**Comment by mattam82** [comment]
Indeed the global flag is not ideal but pretyping doesn't take a record of flags yet, probably that's what should be done. I can see that this change ensures that any original GHole gets an evar created for it and I think this actually solves a bug so I'm not sure it should be conditioned on a flag. E.g pretyping [fun x : ?[T] => ?T) : nat -> Type] should currently raise an error because without this code, the creation of ?[T] is not done and ?T is hence used before it is declared. @herbelin what do you think?

**Comment by ejgallego** [comment]
> pretyping doesn't take a record of flags yet, probably that's what should be done.

@ppedrot has been talking about having an open pretyper for a long time too; I wonder how it does relate here.

But indeed if the flag is hiding some more fundamental problem it'd be great if it could be solved without adding yet another special case.

**Comment by Matafou** [comment]
I think we tested this and it makes the pretyper much slower.

File: lib/flags.mli
Comment (forestjulien) [inline lib/flags.mli]: Allready done in last commit

**Comment by forestjulien** [comment]
@Matafou In fact, we did not test that, I just suspect that remove the flag (and then create the evar systematically) may slow the pretyper.

**Comment by Matafou** [comment]
Indeed this raises an error:

Goal nat -> Type.
pose proof ((fun (x: ?[A]) => ?[A]):nat -> Type).

File: test-suite/bugs/closed/5315.v
Comment (forestjulien) [inline test-suite/bugs/closed/5315.v]: Yes, I'll make a push on that in a few minutes (time needed to recompile) and for your previous comment

Comment (tchajed) [inline test-suite/bugs/closed/5315.v]: Or better still, can the previously working code be left uncommented in the test for a more complete regression test?

**Comment by forestjulien** [comment]
@mattam82: There is a pair of places where evars are not created for holes in pretyper. i can try to expand this solution if you whish.

**Comment by ejgallego** [comment]
If you need a benchmark folks just let me now and I'll launch one in `ci.inria.fr`.

**Comment by Matafou** [comment]
Random tests gave 5-10% slower if we replace if is_function() by true.

**Comment by ejgallego** [comment]
@Matafou maybe it would be interesting to see what true effect on benchmarks this has, I'll submit a bench.

On the other hand, following @mattam82 's suggestion, we should maybe add a (optional) pretyping flags field, however I am not sure about the impact, that start to look more like 8.8 stuff I'd guess.

**Comment by Matafou** [comment]
> @Matafou maybe it would be interesting to see what true effect on benchmarks this has, I'll submit a bench.

Yes that would be good. Don't forget to remove the test if is_function_mode().

Actually the bug found by @mattam82 only happens if there is an explicit evar in the lambda abstraction, we can probably trigger evar generation only when function_mode is on OR when there is an explicit evar.


> On the other hand, following @mattam82 's suggestion, we should maybe add a (optional) pretyping flags field, however I am not sure about the impact, that start to look more like 8.8 stuff I'd guess.

Mmmh indeed removing completely is_program_mode seems difficult and may lead to patch thing here and there. Maybe we can merge this PR (after cleaning) as it is and plan this modification for 8.8.

The thing is that this bug of Function is buging monadic notation (bind (fun X => Y)) so one cannot use such notation with Function. it is quite annoying to let this in 8.7.

**Comment by ejgallego** [comment]
I agree about the problem being important however IMVHO introducing a new global flag in a core component has to be weighted very carefully as it creates many downsides of its own.

Bench running here: https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/155/console

**Comment by Matafou** [comment]
This flag has no concrete syntax, it is only a global variable set when calling the pretyper from plugin indfun. I don't see how it can be really harmful to workers or anything like that.

**Comment by ejgallego** [comment]
@Matafou each new (boolean) global flag multiplies by **2** the number of configurations that have to be tested out of line (given that the change of behavior is not documented in the arguments).

Others may differ, but IMHO if a plugin needs to add imperative flags to the core system in order to work then I don't understand the point of plugins.

**Comment by forestjulien** [comment]
@ejgallego even if I'm not sure to completely agree with @mattam82 point of view (I'm not sure that this patch solve the bug he mentions), one can argue that, in this case, the bug comes from the core system itself. Indeed, I've just put that imperative flag only to ensure that the change is only restricted to the use of Funind plugin but I guess we can globally set the boolean is_function_mode to true  without changing anything in Coq behavior (except maybe time and/or memory consumption)

**Comment by herbelin** [comment]
@mattam82: Indeed, you're right. The hole can define an evar name and hence must be declared. Well spot.

(Apparently, there is another question which is that an evar name stops to be available when the corresponding evar is defined, so that the example that you give apparently does not work even with the change. Something shall have to be done on this side. See also the CEP on [evar names](https://github.com/coq/ceps/pull/23))

Maybe the code of `pretype_type` can actually be simplified by removing the whole `GHole` case. After all, this is (a priori) just a shortcut to avoid creating an evar, not a necessity.

**Comment by forestjulien** [comment]
In made two compilations on the same machine (my laptop) with and without set flag globally sets.
here are the results: 
with function_mode set:

real    7m37.903s
user    23m36.228s
sys     1m52.848s

without function_mode set:

real    7m35.271s
user    23m49.812s
sys     1m54.308s

**Comment by ejgallego** [comment]
So far the differences in the benchmark seem to be 0:

```
┌─────────────────────┬─────────────────────────┬─────────────────────────────────────┬─────────────────────────────────────┐
│                     │      user time [s]      │             CPU cycles              │          CPU instructions           │
│                     │                         │                                     │                                     │
│        package_name │     NEW     OLD PDIFF   │           NEW           OLD PDIFF   │           NEW           OLD PDIFF   │
├─────────────────────┼─────────────────────────┼─────────────────────────────────────┼─────────────────────────────────────┤
│ coq-formal-topology │   36.85   37.37 -1.39 % │  100425031571  100628242597 -0.20 % │  138921607018  138899851053 +0.02 % │
├─────────────────────┼─────────────────────────┼─────────────────────────────────────┼─────────────────────────────────────┤
│           coq-flocq │   52.44   52.47 -0.06 % │  144151001677  143921029028 +0.16 % │  185983808834  185958761050 +0.01 % │
├─────────────────────┼─────────────────────────┼─────────────────────────────────────┼─────────────────────────────────────┤
│  coq-bedrock-facade │ 1070.45 1066.03 +0.41 % │ 2969459065989 2957715078277 +0.40 % │ 4084007757042 4080494369358 +0.09 % │
├─────────────────────┼─────────────────────────┼─────────────────────────────────────┼─────────────────────────────────────┤
│     coq-bedrock-src │ 1050.64 1045.67 +0.48 % │ 2915499226478 2902345433260 +0.45 % │ 4029309171747 4024943599189 +0.11 % │
├─────────────────────┼─────────────────────────┼─────────────────────────────────────┼─────────────────────────────────────┤
│        coq-compcert │  820.59  815.89 +0.58 % │ 2273402147562 2264514939773 +0.39 % │ 3504946713653 3505346485798 -0.01 % │
└─────────────────────┴─────────────────────────┴─────────────────────────────────────┴─────────────────────────────────────┘
```

**Comment by forestjulien** [comment]
@mattam82 @herbelin: I'am not sure what modification you want in pretyping. I suggest either we that PR as it is or one and you will do your modif later or one of you push a modif in my branch.

**Comment by forestjulien** [comment]
concerning @ejgallego test: my feeling after a new read of my modif in pretyping is that, since I just declare a "solved" evar (I declare the variable just before solve with the given solution) there is no real reason for a significant time increasing (maybe a little memory one but not that much)

**Comment by herbelin** [comment]
@forestjulien: I understood that @mattam82 was suggesting that you simply remove the flag, i.e. do as if it were always true, even if not for `Function`.

I was additionally questioning about whether the `GHole` case of `pretype_type` was ever necessarily. It would simplify the code to remove it. It was supposed to be an optimization but it is maybe not worth the burden. Moreover, looking at it again, I'm unsure of the `e_cumul` which would suppose a contravariant subtyping (since `valcon` is not empy only in the `Lambda` case), while CIC implements a non-variant subtyping on the domain of functions.

**Comment by ejgallego** [comment]
Benchmarks show no impact so far (on 8.6 thou) so indeed @mattam82 's suggestion of removing the flag may make sense, or even as @herbelin propose a deeper modification would be welcome. In both case IMHO it would be great to avoid the flag.

However this PR implies a change in the pretyper so it could be better to target it to 8.7 which is almost ready for a beta I understand.

**Comment by forestjulien** [comment]
Ok: 
1- I will push a version without the flag 
2- the @herbelin suggestion seems for me but I would like to know if we won't have any problems with things I don't really understand (like universes and stuff like that) 
3- We can postpone the merge to v8.7 but I don't really see a modification in the pretyper (we just add a new evar and nothing else) as long I we do not change according to @herbelin.

**Comment by ejgallego** [comment]
```
┌─────────────────────┬─────────────────────────┬─────────────────────────────────────┬───────────────────────────────────────┐
│                     │      user time [s]      │             CPU cycles              │           CPU instructions            │
│                     │                         │                                     │                                       │
│        package_name │     NEW     OLD PDIFF   │           NEW           OLD PDIFF   │            NEW            OLD PDIFF   │
├─────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┤
│ coq-formal-topology │   36.85   37.37 -1.39 % │  100425031571  100628242597 -0.20 % │   138921607018   138899851053 +0.02 % │
├─────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┤
│             coq-vst │ 1673.40 1675.02 -0.10 % │ 4652389189830 4654920573486 -0.05 % │  7207007637958  7203839612625 +0.04 % │
├─────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┤
│           coq-flocq │   52.44   52.47 -0.06 % │  144151001677  143921029028 +0.16 % │   185983808834   185958761050 +0.01 % │
├─────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┤
│    coq-fiat-parsers │  701.05  699.47 +0.23 % │ 1920175794158 1919779170510 +0.02 % │  3085755316420  3084758393398 +0.03 % │
├─────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┤
│          coq-geocoq │ 2397.07 2390.76 +0.26 % │ 6676467379166 6659938353567 +0.25 % │ 11364885208326 11359104085532 +0.05 % │
├─────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┤
│  coq-bedrock-facade │ 1070.45 1066.03 +0.41 % │ 2969459065989 2957715078277 +0.40 % │  4084007757042  4080494369358 +0.09 % │
├─────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┤
│     coq-bedrock-src │ 1050.64 1045.67 +0.48 % │ 2915499226478 2902345433260 +0.45 % │  4029309171747  4024943599189 +0.11 % │
├─────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┤
│        coq-compcert │  820.59  815.89 +0.58 % │ 2273402147562 2264514939773 +0.39 % │  3504946713653  3505346485798 -0.01 % │
├─────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┤
│           coq-color │  629.55  625.42 +0.66 % │ 1741869207093 1736697011925 +0.30 % │  2284849158276  2285015011759 -0.01 % │
└─────────────────────┴─────────────────────────┴─────────────────────────────────────┴───────────────────────────────────────┘
```

File: pretyping/pretyping.ml
Comment (ejgallego) [inline pretyping/pretyping.ml]: spurious diff.

File: pretyping/pretyping.ml
Comment (ejgallego) [inline pretyping/pretyping.ml]: spurious diffs too.

File: plugins/funind/glob_term_to_relation.ml
Comment (ejgallego) [inline plugins/funind/glob_term_to_relation.ml]: another spurious nit

**Comment by mattam82** [comment]
@ejgallego your test results are one the branch without the flag?

**Comment by ejgallego** [comment]
> @ejgallego your test results are one the branch without the flag?

@mattam82 the build I launched is @forestjulien 's branch with the flag "always on". I can relaunch a new one if needed.

**Comment by Matafou** [comment]
Do we consider this change small enough for merging? @forestjulien mentioned some other places where an evar is not created. It is not fixed is it? Should we try another test with all these fixed?

@herbelin removing the GHole case completely is maybe a too big change this late? Shall we try just after 8.7 is released?

Does @maximedenes have an opinion on this?

**Comment by maximedenes** [comment]
> Does @maximedenes have an opinion on this?

I would obviously advise against such changes in the pretyper so late in the release cycle. But I'm going to create the branch, so if a fix is agreed on, it can be merged to trunk.

**Comment by Matafou** [comment]
Too bad to let these bug go in 8.7 but I understand. Would it more acceptable if the new behavior was triggered only when inside Function? This was the idea at first (using an ugly global var similar to program_mode).

**Comment by herbelin** [comment]
What about proposing a distinct PR to first "fix" `pretype_type` (assuming an agreement on the fix, a minima, expecting `IntroAnonymous` in the `GHole` branch, or even removing the `GHole` branch)? That would avoid introducing this flag which is after all not needed iiuc.

**Comment by forestjulien** [comment]
> That would avoid introducing this flag which is after all not needed iiuc.

There is no more flag in the last version. 

>  it can be merged to trunk.

The patch is for the v8.6 for now

**Comment by Matafou** [comment]
@herbelin the "fix" is too critical to be integrated in 8.7 as stated by @maximedenes. We will do this on trunk. Not a big deal for the bug detected by @mattam82 because it is a corner case that probably will almost never appear.

But if we do nothing at all the bug in Function remains in 8.7, which is a bit more annoying (e.g. monadic notation cannot be used in Function). That is why I suggest a more delimited fix in 8.7, followed by a real fix in trunk.

**Comment by ejgallego** [comment]
IMO this is a bugfix, has not performance impact, thus if Travis allows it is a clear candidate for 8.7; about 8.6 I don't know; here we see again the problem of having a flow 8.6 -> 8.7.

**Comment by Matafou** [comment]
> IMO this is a bugfix, has not performance impact, thus if Travis allows it is a clear candidate for 8.7;

I tend to agree with you, we just need to convince @maximedenes :-)

**Comment by forestjulien** [comment]
@maximedenes @mattam82 @Matafou @ejgallego : what do we do with this ?

**Comment by ejgallego** [comment]
@forestjulien I understand that this may be too late for 8.6 as 8.6.1 is all but tagged, right @maximedenes ?

As the bugfix doesn't seem to cause any functional or performance regression in 8.7 I guess it should be a candidate for 8.7 , but I guess this is the call of the pretyper maintainers ( @mattam82 , @herbelin ) .

My initial objection (the global flag) is gone. Indeed I think it was a reasonable objection to raise, in most cases such flags are either unneeded or they are masking a deeper design problem IMHO.

File: pretyping/pretyping.ml
Comment (ppedrot) [inline pretyping/pretyping.ml]: May I nitpick after the battle here: a comment explaining why we do that would be worthwhile. It is not explained in the commit message, and my natural reaction by reading this piece of code would be to revert it, as it looks semantically like the identity (except for generating a lone dangling evar).

**Review by ppedrot — CHANGES_REQUESTED** [review]

File: pretyping/pretyping.ml
Comment (forestjulien) [inline pretyping/pretyping.ml]: Your totally right. I'll push a change in a few hours adding the comment.

Comment (Matafou) [inline pretyping/pretyping.ml]: Perhaps a more precise explanation would be adequat.

1) if the lambda was explicitely typed with an explicit meta, we must keep this meta and instantiate it as it may be used later in the term.
2) Several tools need feedback about how implicit types have been instantiated afterwards, thus we need to keep the binding.

**Comment by ejgallego** [comment]
@forestjulien can you rebase this to `master`, other than that, the fix should be ready to go in.

**Comment by forestjulien** [comment]
I'll try do to that during the week-end (I'm not at work today and the connection here is too horrible to do anything but writing comments)

**Comment by forestjulien** [comment]
I would like that someone with a good understanding of pretyping checks this parts before any merge

**Comment by ejgallego** [comment]
So IIUC this is ready pending @ppedrot @mattam82 @herbelin approval.

**Review by ppedrot — APPROVED** [review]
LGTM.

**Comment by forestjulien** [comment]
is the need: backport  means I've to backport the patch to v8.6 ? If so, I would have been aware of that *before*, I had to rebase  it to master -:). Btw, if so must I repush on the same branch or create a new PR ?

**Comment by Zimmi48** [comment]
@forestjulien Don't worry you don't need to care about the `needs: backport` flag. It is there to remind *me* to backport the PR in the v8.7 branch.

**Comment by forestjulien** [comment]
@Zimmi48 : ok but I can do it for you if you whish

**Comment by Zimmi48** [comment]
@forestjulien No need. I've already done it on my fork and I'm waiting on Travis testing to push it.

**Comment by ejgallego** [comment]
@forestjulien current policy is that developers target `master` and RM will pick the worthwhile fixes and backport them.

Thus, I suggest you prepare all your patches against `master`.

**Pull request merged**
